### PR TITLE
Serialize module config

### PR DESCRIPTION
**Pull request by jakemac53**
Uses `JsonSerializable`, so this is a decent e2e test of our own products which is good :). Already resulted in a few bugs being filed.

cc @kevmoo

### CODE REVIEW

File: build_compilers/test/build_test.dart [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  String pkgRoot;
+  try {
+    pkgRoot = _runProc('git', ['rev-parse', '--show-toplevel']);
+    var currentDir = Directory.current.resolveSymbolicLinksSync();
+    if (!p.isWithin(pkgRoot, currentDir)) {
+      throw new StateError('Expected the git root ($pkgRoot) '
+          'to be a parent of the current directory ($currentDir).');
+    }
+  } catch (e) {
+    print("Skipping this test – git didn't run correctly");
```
Comment (natebosch) [inline build_compilers/test/build_test.dart]: [nit] don't mix quote types.

Comment (jakemac53) [reply build_compilers/test/build_test.dart]: This is just because its using a `'` inside the string, which is allowed

Comment (kevmoo) [reply build_compilers/test/build_test.dart]: There's a lint for that 😄

Comment (natebosch) [reply build_compilers/test/build_test.dart]: oops, missed the `'`. Carry on, nothing to see here.

**Pull request merged**